### DISCUSSION TITLE
Making the P2P system modular

### DISCUSSION THREAD

**Discussion by l0k18** [tendermint/tendermint#8361] [Ideas]
I am in the early stages of designing a hybrid blockchain protocol and Tendermint is in the centre of the design for many reasons, but mainly finality. However, as I can see others have also made the observation, the paths by which messages propagate is random, and acts as a structural limit to the latency of messages to the granularity of blocks, which simply isn't fast enough to enable a realtime application.

As I see it, the first step is to generalise and make the P2P package modular, to turn its parts into an interface, and allow a modified propagation system in order to enable changing from the simple raw gossip diffusion model to something more structured.

I am digging through the source code to find where `internal/p2p` touches the rest of the system, and so far I can confirm that most of it is only connected to `internal/` except for the one unit, `node`.

I am putting this issue up for discussion because I will be doing this anyway as a part of my work, and I think that discussions about it amongst the Tendermint developers could be beneficial, both for avoiding easy to avoid roadblocks and to make swapping out the transaction propagation system to a custom one easier, because that is my intention, and I'd rather not have to maintain a whole fork but rather to be able to help make the specific part more loosely coupled so I can just import tendermint and then plug on my custom transaction propagation system.

The proposal, to cut to the chase, is simply to make it possible to swap out the entire p2p package, either by a replace directive, or by some kind of other interface based method. I'm very much in favour of the interface route, as this is the most Go idiomatic way to do this, as well as the easiest to maintain (and everyone here knows that's a core directive of Go).

I am working on a fork at [https://github.com/cybriq/tendermint](https://github.com/cybriq/tendermint) and I am introducing this here to prepare the ground for future changes that will, ideally, build the shortest path from changing the p2p subsystem from internal/monolithic to becoming an optionally swappable component.

Just to explain why I am doing this, in a short summary, I want to make a high availability/high partition resistant consensus that works around the 100ms range of times, or, something that can have potentially 10 or more epochs per tendermint block, in order to enable realtime interactions, such as shared document editing, and realtime games, the goal being to enable multiplayer first person shooters, whose timing precision must at or under 100ms to feel interactive. 

This low latency consensus, having weak consistency, is compensated for by a system of limiting the size of clusters running this consensus and then propagating the data upwards to a global consensus that feeds transactions into the mempool. I have been following tendermint/cosmos development since the early days and I know that the consensus can run as much as 300 validators, based on the random diffusion gossip based p2p transaction propagation strategy. 

The design I have in mind allows this to be dynamically, and leaderlessly be broken down further into small units of 5-12 node clusters who focus on high throughput low latency transaction propagation and cooperate as groups to one or more layers of merging their logs to form a canonical block for the next tendermint block finalization round.

Regarding security, the idea would be to bind the low latency concurrent small clusters to their stake in the validator group in order to attach slashing and jailing functionality to these nodes and exclude proactively exclude misbehaviour from this low latency network while maintaining the low latency.

The important thing as regards to this project is simply in that my proposed first task is to change the p2p system to a form that can have alternative propagation strategies added to chains, while keeping the advantages of tendermint and cosmos IBC for the platforms running it.

**Comment by creachadair** [comment]
[ADR 073](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-073-libp2p.md) provides some useful context about our current plans for the Tendermint P2P implementation. As you've observed, the latest release has "internalized" a lot of the parts, mainly as preparation for a swap to libp2p after v0.36. I suspect libp2p will make some of the experiments you've described here easier, though it will probably be a few months in the future, at least. 

There's always a tradeoff between performance and resiliency against Byzantine network behaviour, and the existing Tendermint design emphasizes the latter. The current implementation definitely leaves some perf on the table (though probably not an asymptotic amount). It would be interesting to see some experimental results with different tradeoffs.

It's relatively easy to speed up network propagation in a federated network, but the obvious approaches also make such a network more vulnerable to corrupt nodes, denial-of-service, and liveness faults. It's a tricky balance. If an application can tolerate low resiliency in exchange for high throughput, it may be better served by federation than consensus.

With networks of "interesting" size (e.g., more than a handful of geographically-diverse nodes), I'd be quite surprised if true realtime performance is practical, but I'd be excited to be proven wrong, and will be interested to hear more of what you discover.  Meanwhile, it sounds like a fork is a good place to experiment, and hopefully some of the libp2p work we're planning to do may simplify your effort a bit.

**Comment by l0k18** [comment]
@creachadair I am not talking about just replacing the mostly dumb propagation system, though with a semi deterministic validator ordering (knowing one block ahead) does make it possible to optimize the p2p system to send the transactions first directly to the next validator. 

What I'm talking about is adding a second layer to the consensus, an early, fast layer, that gets around the problem of factorial message cycles by deliberately layering specific elements for a given session of interactive activity to be focused onto one small group of somewhere around 6-12 nodes per session to enable a minimal time to coherence between the nodes. With all nodes pairing up to compare event orderings in 4 message cycles of gossip, concurrently, being N/2 messages concurrently in the cluster, it is conceivable that they can agree on part of the message ordering and timing within that window fast enough for realtime interaction. 

Keeping the cluster size down and the nodes close together in the cluster is critical, and to enable this, the P2P system I am working on has a baseline consensus for discovering the network topology as a graph of nodes organised to form 3 connections with their nearest peers, and an algorithm for dynamically determining a suitable cluster for a given timing target. The cluster works to determine total ordering of events including forking finding and marking concurrent, unrelated transactions as parallel paths, and making use of some simple elements like transaction type and protocol sequencing to exclude partitioning attacks.

**Comment by l0k18** [comment]
I have completed an initial piece of work that detangled a little part of `internal/p2p` which had got caught inside `node`. 

There is two functions, `createPeerManager` and `createRouter` which pertain to types defined in p2p, and were only stuck in `node` because they shared the use of a type called `closer`. Closer is now living in `lib/closer` as it is shared with several `create*` functions related to other possibly entangled parts that are not important to my task as they don't touch `p2p`; namely `pex`, `mempool`, `evidence` and `consensus`, found in `node/setup/`. 

I get the logic of how they came to be tangled like this, as they are crosscutting, linked to the node as dependencies, but grouped logically by the `create` prefix. This is fine if the components are not really separable from `node` but logically, p2p definitely can be separated, hence the efforts I am making here.

~~I am making a pull request to push this into the develop branch. I have been as careful as possible so that very little other than precisely what is changed is affected so there should be little difficulties with merging.~~

edit: I am thinking it is a bit premature to add these changes even though they are very simple and logical. After seeing what is going on here with the somewhat messy tangle between internal and node elements and the rest of the whole thing, I will wait on more feedback about where to go with it. The changes that are current to now, https://github.com/cybriq/tendermint/commit/6ea78c4f720f4a7a7bd67d3210278b9b459f8bb9 are only small refactorings that break a tight coupling between `internal/p2p/` and `node/`. My next step will be to isolate the contact surface between `internal/p2p/` and create an exported encapsulation of it in `libs/` most likely. The idea would be that a fork can then be created which keeps everything else but replaces `libs/p2p/` with an implementation that uses a different approach.

Having read through the ADR 073 document I can see that I am already doing a small part of planned changes already.

**Comment by tychoish** [comment]
>  I am thinking it is a bit premature to add these changes even though they are very simple and logical. After seeing what is going on here with the somewhat messy tangle between internal and node elements and the rest of the whole thing.

I think that's true. Both that the components are initialized in a way that's inideal, and that it's a bit premature to start untagling. My view at the moment is that the peer manager will (mostly? entirely?) disappear in the post-libp2p world, and in light of that I've decided to leave it where it is for the 0.36 cycle and hope that we can solve the peer manager (and some of the initialization) entanglement via deletion rather than refactoring as such.
